### PR TITLE
Dns over http2 5773 v16

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/5773

Describe changes:
- analyze DNS over HTTP2

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1970

https://github.com/OISF/suricata/pull/11428 with DNS v3 logging, no longer a draft, cf last commit

Should there be a squash up of commits ?

@jasonish here for a DOH2 tx, we log a bidirectional HTTP2 transaction, and then if any, a DNS transaction, preferring the answer... What do you think about it ? This allows to keep the same format as for regular dns.
Another option would be to log two doh2 events : one for the DNS request and one for the DNS answer, with HTTP2 getting logged twice... not sure how it would work out for alerts...

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 21444

**Comment by jasonish** [comment]
> @jasonish here for a DOH2 tx, we log a bidirectional HTTP2 transaction, and then if any, a DNS transaction, preferring the answer... What do you think about it ? This allows to keep the same format as for regular dns.

Probably fine, as these are `doh` events, not `dns` event.  Some thoughts tho:
- will there be a `doh3` at some point and so on?
- DNS is valuable metadata, I wonder if spreading it over multiple event types will limit its usefulness. My head already hurts thinking about creating DNS reports that cover tradition DNS and DoH.

**Comment by victorjulien** [comment]
In dcerpc, we don't care if it comes over smb or udp/tcp directly, right? I feel dns records shouldn't be different regardless of their transport. It should be recognizable somehow as Dns over HTTP/2, Dns over HTTP, etc still, but that can be done in various ways.

**Comment by catenacyber** [comment]
So, you would want `dns` event instead of `doh2`
We know it is doh2 because this event has `http` object in it

**Comment by victorjulien** [comment]
> So, you would want `dns` event instead of `doh2` We know it is doh2 because this event has `http` object in it

Yeah that sounds good.

File: rust/src/dns/dns.rs [hunk 1]
```diff
@@ -338,18 +343,116 @@ impl State<DNSTransaction> for DNSState {
...
+    if let Ok((body, header)) = parser::dns_parse_header(input) {
+        if probe_header_validity(&header, input.len()).0 {
+            return Some((body, header));
+        }
+    }
+    None
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum DNSParseError {
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/dns/dns.rs:356]: ok

**Comment by catenacyber** [comment]
Thanks

Continued in https://github.com/OISF/suricata/pull/11498

**Pull request closed without merging**